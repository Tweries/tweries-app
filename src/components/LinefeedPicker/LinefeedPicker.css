.LinefeedPicker__li {
  align-items: center;
  display: flex;
  flex-direction: row;
  margin: 0;
}

.LinefeedPicker__linefeed {
  text-align: center;
}

.LinefeedPicker__ul {
  display: flex;
}
