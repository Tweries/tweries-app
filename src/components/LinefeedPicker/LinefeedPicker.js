import React, { useState } from 'react';
import { LINEFEED, NEWLINE } from '../../constants';

const copy = {
  'Newline(s)': 'Newline(s)'
};

const types = {
  custom: 'custom',
  newline: 'newline'
};

function Custom({ disabled, onChange, value }) {
  return (
    <input
      className="bg-gray-200 border border-gray-500 rounded text-center"
      data-testid="linefeed"
      disabled={disabled}
      maxLength="4"
      onChange={onChange}
      placeholder={LINEFEED}
      size="4"
      value={value}
    />
  );
}

function LinefeedPicker({ feature, onChange }) {
  const [type, setType] = useState(types.newline);
  const [custom, setCustom] = useState(LINEFEED);

  return (
    <ul className="flex">
      <li>
        <input
          className="ml-1"
          data-testid="newline"
          defaultChecked
          name="linefeed"
          onChange={() => {
            setType(types.newline);
            onChange(NEWLINE);
          }}
          type="radio"
          value="newline"
        />
        <label className="ml-1" htmlFor="newline">
          {copy['Newline(s)']}
        </label>
      </li>
      <li>
        <input
          className="ml-1"
          data-testid="custom"
          name="linefeed"
          onChange={() => {
            setType(types.custom);
            onChange(custom);
          }}
          type="radio"
          value="custom"
        />
        <label className="ml-1" htmlFor="louie">
          <Custom
            disabled={type !== types.custom}
            onChange={e => {
              const {
                target: { value }
              } = e;
              setCustom(value);
              onChange(value);
            }}
            value={custom}
          />
        </label>
      </li>
    </ul>
  );
}

export default LinefeedPicker;
