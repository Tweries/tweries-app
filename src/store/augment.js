function augment(reducer) {
  function logAction(state, action) {
    // console.log('state:', state);
    // console.log('action:', action);
    const newState = reducer(state, action);
    // console.log('newState:', newState);
    return newState;
  }
  return logAction;
}

export default augment;
