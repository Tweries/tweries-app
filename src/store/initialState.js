const HASHTAGS = '#QueenFTW';
const SOURCE = `Mama, just killed a man
Put a gun against his head
Pulled my trigger, now he's dead
Mama, life had just begun
But now I've gone and thrown it all away
[..]
Mama, ooh
Didn't mean to make you cry
If I'm not back again this time tomorrow
Carry on, carry on
As if nothing really matters
[..]
Too late, my time has come
Sends shivers down my spine
Body's aching all the time
Goodbye everybody, I've got to go
Gotta leave you all behind and face the truth`;

const initialState = {
  hashtags: HASHTAGS,
  healthy: false,
  items: [],
  source: SOURCE,
  userId: null
};

export default initialState;
