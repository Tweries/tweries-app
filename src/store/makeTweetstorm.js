import v from 'voca';

const LINEFEED = '[..]';
const MAX_LENGTH = 280;
const PREFIX_PLACEHOLDER = '_/_';

function makeSequenceNumber(index, length) {
  if (index === undefined && length === undefined) {
    return PREFIX_PLACEHOLDER;
  }
  return `${index + 1}/${length}`;
}

function makeTweetstorm(source, hashtags) {
  let copy = source.slice();
  const parts = [];

  while (copy.length !== 0) {
    let take = v.prune(
      copy,
      MAX_LENGTH - makeSequenceNumber().length - hashtags.length - 2, // INFO: 1 space after the prefix and one space before the suffix
      ''
    );
    if (take.indexOf(LINEFEED) > -1) {
      take = v.substr(take, 0, take.indexOf(LINEFEED));
      copy = v.substr(copy, take.length + 4); // INFO: 4 is the lenght of the linefeed
    } else {
      copy = v.substr(copy, take.length + 1); // INFO: 1 is the space after the word
    }
    parts.push(take);
  }

  const tweetstorm = parts.map((part, index) => {
    const tweet = `${v.trim(part)} ${hashtags} ${makeSequenceNumber(
      index,
      parts.length
    )}`;
    return { length: tweet.length, tweet };
  });

  return tweetstorm;
}

export default makeTweetstorm;
