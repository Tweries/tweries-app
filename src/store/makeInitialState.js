import { NEWLINE } from '../constants';
import makeTweetstorm from './makeTweetstorm';

function makeInitialState({ feature }) {
  return {
    healthy: false,
    items: makeTweetstorm(feature)({
      linefeed: NEWLINE,
      source: ''
    }),
    linefeed: NEWLINE,
    notification: null,
    source: '',
    userId: ''
  };
}

export default makeInitialState;
