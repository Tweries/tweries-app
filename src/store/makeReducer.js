import { SUCCESS } from '../constants';
import makeInitialState from './makeInitialState';
import makeTweetstorm from './makeTweetstorm';

export const types = {
  APPEND_SCREEN_NAME: 'APPEND_SCREEN_NAME',
  CHANGE_SOURCE: 'CHANGE_SOURCE',
  CHANGE_TWEET: 'CHANGE_TWEET',
  DISMISS_TOAST: 'DISMISS_TOAST',
  RESET_TWEETSTORM: 'RESET_TWEETSTORM',
  SET_HEALTHY: 'SET_HEALTHY',
  SET_USER_ID: 'SET_USER_ID'
};

function makeReducer(feature) {
  function reducer(state = makeInitialState({ feature }), action) {
    switch (action.type) {
      case types.APPEND_SCREEN_NAME: {
        const source = `@${action.value} ${state.source}`;
        return {
          ...state,
          items: makeTweetstorm(feature)({
            linefeed: state.linefeed,
            source
          }),
          source
        };
      }
      case types.CHANGE_SOURCE: {
        return {
          ...state,
          items: makeTweetstorm(feature)({
            linefeed: state.linefeed,
            source: action.value
          }),
          source: action.value
        };
      }
      case types.CHANGE_TWEET: {
        return {
          ...state,
          items: [...state.items].map((item) => {
            const copy = { ...item };
            if (copy.id === action.value.id) {
              copy.tweet = action.value.tweet;
            }
            return copy;
          })
        };
      }
      case types.DISMISS_TOAST: {
        return {
          ...state,
          notification: null
        };
      }
      case types.RESET_TWEETSTORM: {
        if (action.value.type === SUCCESS) {
          return {
            ...makeInitialState({ feature }),
            healthy: state.healthy,
            notification: action.value,
            userId: state.userId
          };
        }
        return {
          ...state,
          notification: action.value
        };
      }
      case types.SET_HEALTHY: {
        return {
          ...state,
          healthy: action.value
        };
      }
      case types.SET_USER_ID: {
        return {
          ...state,
          userId: action.value
        };
      }
      default:
        return state;
    }
  }

  return reducer;
}

export default makeReducer;
