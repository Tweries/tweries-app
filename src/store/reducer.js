import makeInitialState from './makeInitialState';
import makeTweetstorm from './makeTweetstorm';

export const types = {
  CHANGE_HASHTAGS: 'CHANGE_HASHTAGS',
  CHANGE_LINEFEED: 'CHANGE_LINEFEED',
  CHANGE_SOURCE: 'CHANGE_SOURCE',
  RESET_TWEETSTORM: 'RESET_TWEETSTORM',
  SET_HEALTHY: 'SET_HEALTHY',
  SET_USER_ID: 'SET_USER_ID'
};

function reducer(state = makeInitialState({}), action) {
  switch (action.type) {
    case types.CHANGE_HASHTAGS: {
      return {
        ...state,
        hashtags: action.value,
        items: makeTweetstorm({
          feature: state.feature,
          hashtags: action.value,
          linefeed: state.linefeed,
          source: state.source
        })
      };
    }
    case types.CHANGE_LINEFEED: {
      return {
        ...state,
        linefeed: action.value,
        items: makeTweetstorm({
          feature: state.feature,
          hashtags: state.hashtags,
          linefeed: action.value,
          source: state.source
        })
      };
    }
    case types.CHANGE_SOURCE: {
      return {
        ...state,
        items: makeTweetstorm({
          feature: state.feature,
          hashtags: state.hashtags,
          linefeed: state.linefeed,
          source: action.value
        }),
        source: action.value
      };
    }
    case types.RESET_TWEETSTORM: {
      return {
        ...makeInitialState({}),
        healthy: state.healthy,
        userId: state.userId
      };
    }
    case types.SET_HEALTHY: {
      return {
        ...state,
        healthy: action.value
      };
    }
    case types.SET_USER_ID: {
      return {
        ...state,
        userId: action.value
      };
    }
    default:
      return state;
  }
}

export default reducer;
