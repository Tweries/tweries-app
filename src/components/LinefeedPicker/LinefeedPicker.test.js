import { render, fireEvent } from '@testing-library/react';
import React from 'react';
import { LINEFEED } from '../../store/makeTweetstorm';
import {
  PICK_YOUR_OWN_LINEFEED_V1,
  PICK_YOUR_OWN_LINEFEED_V2
} from '../../feature';
import LinefeedPicker from './LinefeedPicker';

describe('LinefeedPicker', () => {
  it('simple', () => {
    const { container } = render(
      <LinefeedPicker feature={{ active: () => false }} />
    );

    expect(container).toMatchSnapshot();
  });

  it('v1', () => {
    const mockOnChange = jest.fn();
    const { getByTestId } = render(
      <LinefeedPicker
        feature={{
          active: feature => feature === PICK_YOUR_OWN_LINEFEED_V1
        }}
        onChange={mockOnChange}
      />
    );

    fireEvent.change(getByTestId('linefeed'), { target: { value: '||' } });

    expect(mockOnChange).toBeCalled();
  });

  it('v2', () => {
    const mockOnChange = jest.fn();
    const { getByTestId } = render(
      <LinefeedPicker
        feature={{
          active: feature => feature === PICK_YOUR_OWN_LINEFEED_V2
        }}
        onChange={mockOnChange}
      />
    );

    fireEvent.click(getByTestId('custom'));
    fireEvent.change(getByTestId('linefeed'), { target: { value: '||' } });
    fireEvent.click(getByTestId('newline'));

    expect(mockOnChange.mock.calls).toMatchSnapshot();
  });
});
