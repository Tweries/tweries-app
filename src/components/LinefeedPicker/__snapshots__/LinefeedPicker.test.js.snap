// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinefeedPicker simple 1`] = `
<div>
  <span>
    ;;
  </span>
</div>
`;

exports[`LinefeedPicker v2 1`] = `
Array [
  Array [
    ";;",
  ],
  Array [
    "||",
  ],
  Array [
    "
",
  ],
]
`;
